# HG changeset patch
# User Aaron Schrab <aaron@schrab.com>
# Date 1454288688 0
#      Mon Feb 01 01:04:48 2016 +0000
# Branch gentoo-1.7
# Node ID a0c62e12a105147bc1d540ab7b4940cff309803f
# Parent  8eac2e5c1219127e1bf85880dfd15370f36179f8
[PATCH] feature: cond-date
From c935627f21a210dcf7f874969655e728891c79d5 Mon Sep 17 00:00:00 2001
Use rules to choose date format
---
 hdrline.c | 92 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 init.h    |  4 +++
 muttlib.c | 16 ++++++++---
 3 files changed, 108 insertions(+), 4 deletions(-)

diff -r 8eac2e5c1219 -r a0c62e12a105 PATCHES
--- a/PATCHES	Mon Feb 01 00:41:48 2016 +0000
+++ b/PATCHES	Mon Feb 01 01:04:48 2016 +0000
@@ -1,3 +1,4 @@
+patch-cond-date-neomutt
 patch-nested-if-neomutt
 patch-initials-neomutt
 patch-ifdef-neomutt
diff -r 8eac2e5c1219 -r a0c62e12a105 hdrline.c
--- a/hdrline.c	Mon Feb 01 00:41:48 2016 +0000
+++ b/hdrline.c	Mon Feb 01 01:04:48 2016 +0000
@@ -329,6 +329,98 @@ hdr_format_str (char *dest,
 	const char *cp;
 	struct tm *tm; 
 	time_t T;
+	int i = 0, invert = 0;
+
+	if (optional && ((op == '[') || (op == '('))) {
+	  char *is;
+	  T = time (NULL);
+	  tm = localtime (&T);
+	  T -= (op == '(') ? hdr->received : hdr->date_sent;
+
+	  is = (char *) prefix;
+	  if (*is == '>') {
+	    invert = 1;
+	    is++;
+	  }
+
+	  while (*is && (*is != '?')) {
+	    int t = strtol (is, &is, 10);
+	    /* semi-broken (assuming 30 days in all months) */
+	    switch (*(is++)) {
+	      case 'y':
+		if (t > 1) {
+		  t--;
+		  t *= (60 * 60 * 24 * 365);
+		}
+		t += ((tm->tm_mon  * 60 * 60 * 24 * 30) +
+		      (tm->tm_mday * 60 * 60 * 24) +
+		      (tm->tm_hour * 60 * 60) +
+		      (tm->tm_min  * 60) +
+		       tm->tm_sec);
+		break;
+
+	      case 'm':
+		if (t > 1) {
+		  t--;
+		  t *= (60 * 60 * 24 * 30);
+		}
+		t += ((tm->tm_mday * 60 * 60 * 24) +
+		      (tm->tm_hour * 60 * 60) +
+		      (tm->tm_min  * 60) +
+		      tm->tm_sec);
+		break;
+
+	      case 'w':
+		if (t > 1) {
+		  t--;
+		  t *= (60 * 60 * 24 * 7);
+		}
+		t += ((tm->tm_wday * 60 * 60 * 24) +
+		      (tm->tm_hour * 60 * 60) +
+		      (tm->tm_min  * 60) +
+		       tm->tm_sec);
+		break;
+
+	      case 'd':
+		if (t > 1) {
+		  t--;
+		  t *= (60 * 60 * 24);
+		}
+		t += ((tm->tm_hour * 60 * 60) +
+		      (tm->tm_min  * 60) +
+		       tm->tm_sec);
+		break;
+
+	      case 'H':
+		if (t > 1) {
+		  t--;
+		  t *= (60 * 60);
+		}
+		t += ((tm->tm_min * 60) +
+		       tm->tm_sec);
+		break;
+
+	      case 'M':
+		if (t > 1) {
+		  t--;
+		  t *= (60);
+		}
+		t += (tm->tm_sec);
+		break;
+
+	      default:
+		break;
+	    }
+	    i += t;
+	  }
+
+	  if (i < 0)
+	    i *= -1;
+
+	  if (((T > i) || (T < (-1*i))) ^ invert)
+	    optional = 0;
+	  break;
+	}
 
 	p = dest;
 
diff -r 8eac2e5c1219 -r a0c62e12a105 init.h
--- a/init.h	Mon Feb 01 00:41:48 2016 +0000
+++ b/init.h	Mon Feb 01 01:04:48 2016 +0000
@@ -1372,6 +1372,10 @@ struct option_t MuttVars[] = {
   ** .dt %*X    .dd soft-fill with character ``X'' as pad
   ** .de
   ** .pp
+  ** Date format expressions can be constructed based on relative dates. Using
+  ** the date formatting operators along with nested conditionals, the date
+  ** format can be modified based on how old a message is.  See the section on
+  ** ``Conditional Dates'' for an explanation and examples
   ** ``Soft-fill'' deserves some explanation: Normal right-justification
   ** will print everything to the left of the ``%>'', displaying padding and
   ** whatever lies to the right only if there's room. By contrast,
diff -r 8eac2e5c1219 -r a0c62e12a105 muttlib.c
--- a/muttlib.c	Mon Feb 01 00:41:48 2016 +0000
+++ b/muttlib.c	Mon Feb 01 01:04:48 2016 +0000
@@ -1235,7 +1235,15 @@ void mutt_FormatString (char *dest,		/* 
       if (*src == '<')
       {
 	flags |= MUTT_FORMAT_OPTIONAL;
+	ch = *(++src); /* save the character to switch on */
 	src++;
+	cp = prefix;
+	count = 0;
+	while ((count < sizeof (prefix)) && (*src != '?')) {
+	  *cp++ = *src++;
+	  count++;
+	}
+	*cp = 0;
       }
       else
       {
@@ -1251,12 +1259,12 @@ void mutt_FormatString (char *dest,		/* 
 	  count++;
 	}
 	*cp = 0;
-      }
 
-      if (!*src)
-	break; /* bad format */
+	if (!*src)
+	  break; /* bad format */
 
-      ch = *src++; /* save the character to switch on */
+	ch = *src++; /* save the character to switch on */
+      }
 
       if (flags & MUTT_FORMAT_OPTIONAL)
       {
